### PR TITLE
JDK-8265298: Hard VM crash when deadlock between "access" and higher ranked lock is detected

### PR DESCRIPTION
**Pull request by shipilev**
I stumbled upon this when doing some Shenandoah work. The development code tried to lock the `leaf` lock, while already holding the `access` lock. Normally it would have been detected by VM, but instead, we tried to recursively acquire `tty_lock` for `Thread::print_owned_locks`. But that `tty` lock is still ranked higher than `access`, so deadlock detection triggers over and over again until we run out of stack and crash hard. `tty` and `access` are ranked this way because of [JDK-8214315](https://bugs.openjdk.java.net/browse/JDK-8214315). Read the rest in the bug.  

I believe the way out is to only enter `Thread::print_owned_locks` when we know deadlock detection code would not run in circles. New test for `access` and `leaf` shows the original failure. Another new test checks `tty` and `special` to verify that the check should be `> tty`, not `>= tty`.

Additional testing:
 - [x] Linux x86_64 fastdebug tier1
 - [x] New regression tests (fail without the patch, pass wit it)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265298](https://bugs.openjdk.java.net/browse/JDK-8265298): Hard VM crash when deadlock between "access" and higher ranked lock is detected

### Reviewers
 * [Patricio Chilano Mateo](https://openjdk.java.net/census#pchilanomate) (@pchilano - Committer) ⚠️ Review applies to ecf6e5bb8bec49ea573715dc4143f425169e576c
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3524/head:pull/3524` \
`$ git checkout pull/3524`

Update a local copy of the PR: \
`$ git checkout pull/3524` \
`$ git pull https://git.openjdk.java.net/jdk pull/3524/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3524`

View PR using the GUI difftool: \
`$ git pr show -t 3524`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3524.diff">https://git.openjdk.java.net/jdk/pull/3524.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/gtest/runtime/test_mutex_rank.cpp [hunk 1]
```diff
@@ -104,6 +104,36 @@ TEST_VM_ASSERT_MSG(MutexRank, mutex_trylock_rank_out_of_orderB,
   mutex_rankA->unlock();
 }
 
+TEST_VM_ASSERT_MSG(MutexRank, mutex_wait_access_leaf,
```
Comment (pchilano) [inline test/hotspot/gtest/runtime/test_mutex_rank.cpp:107]: s/mutex_wait/mutex_lock

File: test/hotspot/gtest/runtime/test_mutex_rank.cpp [hunk 2]
```diff
@@ -104,6 +104,36 @@ TEST_VM_ASSERT_MSG(MutexRank, mutex_trylock_rank_out_of_orderB,
...
+  Mutex* mutex_rank_access = new Mutex(Mutex::access, "mutex_rank_access", false, Mutex::_safepoint_check_never);
+  Mutex* mutex_rank_leaf = new Mutex(Mutex::leaf, "mutex_rank_leaf", false, Mutex::_safepoint_check_never);
+
+  mutex_rank_access->lock_without_safepoint_check();
+  mutex_rank_leaf->lock_without_safepoint_check();
+  mutex_rank_leaf->unlock();
+  mutex_rank_access->unlock();
+}
+
+TEST_VM_ASSERT_MSG(MutexRank, mutex_wait_tty_special,
```
Comment (pchilano) [inline test/hotspot/gtest/runtime/test_mutex_rank.cpp:122]: s/mutex_wait/mutex_lock

File: test/hotspot/gtest/runtime/test_mutex_rank.cpp [hunk 3]
```diff
@@ -165,4 +195,36 @@ TEST_VM_ASSERT_MSG(MutexRank, monitor_wait_rank_special,
...
+                   "-- possible deadlock") {
+  JavaThread* THREAD = JavaThread::current();
+  ThreadInVMfromNative invm(THREAD);
+
+  Monitor* monitor_rank_tty = new Monitor(Mutex::tty, "monitor_rank_tty", false, Mutex::_safepoint_check_never);
+  Monitor* monitor_rank_special = new Monitor(Mutex::special, "monitor_rank_special", false, Mutex::_safepoint_check_never);
+
+  monitor_rank_tty->lock_without_safepoint_check();
+  monitor_rank_special->lock_without_safepoint_check();
+  monitor_rank_special->wait_without_safepoint_check(1);
```
Comment (pchilano) [inline test/hotspot/gtest/runtime/test_mutex_rank.cpp:226]: Same as above, should lock special first and then tty one.

File: test/hotspot/gtest/runtime/test_mutex_rank.cpp [hunk 4]
```diff
@@ -165,4 +195,36 @@ TEST_VM_ASSERT_MSG(MutexRank, monitor_wait_rank_special,
...
+                   "-- possible deadlock") {
+  JavaThread* THREAD = JavaThread::current();
+  ThreadInVMfromNative invm(THREAD);
+
+  Monitor* monitor_rank_access = new Monitor(Mutex::access, "monitor_rank_access", false, Mutex::_safepoint_check_never);
+  Monitor* monitor_rank_leaf = new Monitor(Mutex::leaf, "monitor_rank_leaf", false, Mutex::_safepoint_check_never);
+
+  monitor_rank_access->lock_without_safepoint_check();
+  monitor_rank_leaf->lock_without_safepoint_check();
+  monitor_rank_leaf->wait_without_safepoint_check(1);
```
Comment (pchilano) [inline test/hotspot/gtest/runtime/test_mutex_rank.cpp:210]: If you want to exercise the wait() error case you should lock the leaf lock first and then the access one. Then you will get the assert at wait_without_safepoint_check(1). You will have to fix the error message that we check for above to be the one used in the wait case.

**Review by pchilano — APPROVED** [review]
Hi Aleksey,

Fix looks good to me. We could even avoid that print altogether since we already have the offending locks in the assert message, and the hs_err file will show the owners of all locks anyways. But I guess it doesn't hurt to try and print that when we know is safe.
Only comments on the added test cases.

Thanks,
Patricio

File: test/hotspot/gtest/runtime/test_mutex_rank.cpp (hunk 3 above)
Comment (pchilano) [inline test/hotspot/gtest/runtime/test_mutex_rank.cpp:226]: Same as above, should lock special first and then tty one.

File: test/hotspot/gtest/runtime/test_mutex_rank.cpp (hunk 4 above)
Comment (shipilev) [reply test/hotspot/gtest/runtime/test_mutex_rank.cpp:210]: The intent is to lock `access` first, then `leaf` -- after all, that was the issue I initially found. About `wait`: I copy-pasted the shape of the test from the `monitor_wait_rank_special` above. I thought it was dubious that we basically expect to assert at `monitor_rank_leaf->lock_without_safepoint_check();`, but thought having `wait` makes the code a bit more readable. I can just remove `wait` here.

Comment (shipilev) [reply test/hotspot/gtest/runtime/test_mutex_rank.cpp:210]: Ah, you mean the test checks the attempt to wait! Right, let me fix that.

File: test/hotspot/gtest/runtime/test_mutex_rank.cpp (hunk 1 above)
Comment (shipilev) [reply test/hotspot/gtest/runtime/test_mutex_rank.cpp:107]: Renamed.

File: src/hotspot/share/runtime/mutex.cpp [hunk 5]
```diff
@@ -412,7 +412,12 @@ void Mutex::check_rank(Thread* thread) {
     // to acquire, then deadlock prevention rules require that the rank
     // of m2 be less than the rank of m1. This prevents circular waits.
     if (least != NULL && least->rank() <= this->rank()) {
-      thread->print_owned_locks();
+      if (least->rank() > Mutex::tty) {
+        // Printing owned locks acquires tty lock. If lowest rank is below or equal
```
Comment (navyxliu) [inline src/hotspot/share/runtime/mutex.cpp:416]: If the least* rank was* below or equal tty,

File: test/hotspot/gtest/runtime/test_mutex_rank.cpp (hunk 1 above)
Comment (navyxliu) [reply test/hotspot/gtest/runtime/test_mutex_rank.cpp:107]: +1

File: test/hotspot/gtest/runtime/test_mutex_rank.cpp (hunk 4 above)
Comment (shipilev) [reply test/hotspot/gtest/runtime/test_mutex_rank.cpp:210]: Fixed. There is a wrinkle: `leaf` is not a `special` lock, so this test would make little sense. I replaced `leaf` with `tty` instead.

File: test/hotspot/gtest/runtime/test_mutex_rank.cpp (hunk 3 above)
Comment (shipilev) [reply test/hotspot/gtest/runtime/test_mutex_rank.cpp:226]: Fixed.

**Comment by navyxliu** [comment]
hi, Aleksey, 

Thank you for fixing this issue. I ran into it too!
Your fix looks reasonable to me.

File: src/hotspot/share/runtime/mutex.cpp (hunk 5 above)
Comment (shipilev) [reply src/hotspot/share/runtime/mutex.cpp]: Fixed.

File: test/hotspot/gtest/runtime/test_mutex_rank.cpp (hunk 4 above)
Comment (pchilano) [reply test/hotspot/gtest/runtime/test_mutex_rank.cpp]: Fix looks good. I thought you wanted to check the other failure mode in wait() (the one where you have to wait on the least rank owned monitor).

**Comment by pchilano** [comment]
Thanks Aleksey, fixes look good!

Patricio

**Review by dholmes-ora — APPROVED** [review]
Looks good.

Thanks,
David

**Pull request closed without merging**